packages/notte-core/src/notte_core/actions/typedicts.py

```
@@ -340,3 +341,32 @@ def action_dict_to_base_action(data: ActionDict) -> "BaseAction":
+def parse_action(action: "BaseAction | None" = None, **kwargs: Any) -> "BaseAction | ActionUnion":
+    from notte_core.actions.actions import (
+        ActionValidation,
+        BaseAction,
+    )
+
+    # Fast path: if action is already a BaseAction, use it directly
+    if isinstance(action, BaseAction):
+        step_action = action
+    elif kwargs:
+        if "type" not in kwargs:
+            raise ValueError("Missing required action field: 'type'")
```

Ruff (0.14.10)

Avoid specifying long messages outside the exception class

(TRY003)